# MHPA analysis pipeline
 
This repository contains an in-house computational pipeline developed in Python/Matlab for identification of low frequency mutations in TSC2 and TP53 from Massively Parallel Sequencing (MPS) data generated using the Multiplex High-sensitivity PCR Assay (MHPA) strategy (Klonowska et al, Journal of Clinical Investigation, 2022).
 
The MHPA pipeline has been developed in Dr. David J Kwiatkowski lab (Brigham and Womenâ€™s Hospital, Harvard Medical School, Boston) and is derived from the custom code pipeline used previously in the Kwiatkowski lab for the identification of low frequency allele frequency mutations in TSC1/TSC2 from targeted capture MPS data, which was reported in:
 
*	Giannikou K, et al. Low-level mosaicism in tuberous sclerosis complex: prevalence, clinical features, and risk of disease transmission. Genet Med. 2019;21(11):2639-2643.
* Treichel AM, et al. Phenotypic distinctions between mosaic forms of tuberous sclerosis complex. Genet Med. 2019;21(11):2594-2604.
* Ogorek B, et al. TSC2 pathogenic variants are predictive of severe clinical manifestations in TSC infants: results of the EPISTOP study. Genet Med. 2020;22(9):1489-1497.

For citation of this custom MHPA pipeline code, please use:
Klonowska K, Grevelink JM, Giannikou K, Ogorek B, Herbert ZT, Thorner AR, Darling TN, Moss J, Kwiatkowski DJ, Ultrasensitive profiling of UV mutations identifies thousands of subclinical facial tumors in Tuberous Sclerosis Complex, Journal of Clinical Investigation, 2022, in press.

